File: leetcode/copy_list_with_random_pointer.js
/* A linked list is given such that each node contains an additional random 
 * pointer which could point to any node in the list or null.
 * Return a deep copy of the list.
 * Time Complexity: O(n)
 * Space Complexity: O(1)
 */

/**
 * Definition for singly-linked list with a random pointer.
 * function RandomListNode(label) {
 *     this.label = label;
 *     this.next = this.random = null;
 * }
 */

/**
 * @param {RandomListNode} head
 * @return {RandomListNode}
 */
var copyRandomList = function(head) {
    if (!head)
        return null;
    for (var cur = head; cur; cur = next) {
        var next = cur.next;
        cur.next = new RandomListNode(cur.label);
        cur.next.next = next;
    }
    for (cur = head; cur; cur = cur.next.next) {
        if (cur.random)
            cur.next.random = cur.random.next;
    }
    var result = head.next;
    for (var cur1 = head, cur2 = head.next; cur1; cur1 = cur1.next, cur2 = cur2.next) {
        cur1.next = cur1.next.next;
        cur2.next = cur1.next ? cur2.next.next : null;
    }
    return result;
};

File: lintcode/copy_list_with_random_pointer.py
# Definition for singly-linked list with a random pointer.
# class RandomListNode:
#     def __init__(self, x):
#         self.label = x
#         self.next = None
#         self.random = None
class Solution:
    # @param head: A RandomListNode
    # @return: A RandomListNode
    def copyRandomList(self, head):
        if not head: return None
        cur = head
        while cur:
            cur.next, next = RandomListNode(cur.label), cur.next
            cur.next.next, cur = next, next
        cur = head
        while cur:
            if cur.random: cur.next.random = cur.random.next
            cur = cur.next.next
        cur1, cur2, result = head, head.next, head.next
        while cur1:
            cur1.next = cur1.next.next
            cur2.next = cur2.next.next if cur1.next else None
            cur1, cur2 = cur1.next, cur2.next
        return result
